### PR TITLE
[ncp] update pass through filter rules

### PR DESCRIPTION
**Pull request by librasungirl**
Update NCP filtering rules to exclude:
1. RLOC or ALOC as Destination Address
2. MLE messages
3. TMF messages

### CODE REVIEW

File: src/core/net/ip6.cpp [hunk 1]
```diff
@@ -594,17 +592,34 @@ otError Ip6::ProcessReceiveCallback(const Message &aMessage, const MessageInfo &
...
+                if (aMessageInfo.GetSockAddr().IsLinkLocal() ||
+                    aMessageInfo.GetSockAddr().IsLinkLocalMulticast())
+                {
+                    ExitNow(error = OT_ERROR_NO_ROUTE);
+                }
+            }
+            // do not pass TMF messages
+            else if (udp.GetDestinationPort() == kCoapUdpPort)
+            {
+                if (((GetInstance().mThreadNetif.GetMle().IsMeshLocalAddress(aMessageInfo.GetSockAddr()) ||
```
Comment (jwhui) [inline src/core/net/ip6.cpp]: This duplicates the code from [src/core/thread/thread_netif.cpp](https://github.com/openthread/openthread/blob/master/src/core/thread/thread_netif.cpp#L183).  Can we remove this duplication?

File: src/core/thread/thread_netif.cpp [hunk 2]
```diff
@@ -180,22 +179,24 @@ otError ThreadNetif::RouteLookup(const Ip6::Address &source, const Ip6::Address
     return error;
 }
 
-otError ThreadNetif::TmfFilter(const Message &aMessage, const Ip6::MessageInfo &aMessageInfo, void *aContext)
+otError ThreadNetif::ApplyTmfFilter(const Ip6::MessageInfo &aMessageInfo, Mle::MleRouter &aMle)
```
Comment (bukepo) [inline src/core/thread/thread_netif.cpp]: Is it possible to involve a public non-static method like 
```cpp
bool IsTmfMessage(const Ip6::MessageInfo &aMessageInfo) const
```
and calling it both in `TmfFilter()` and IP layer's filtering function?
It looks strange passing `Mle` as parameter into a function of `ThreadNetif` which is actually owner of the `Mle`.

File: src/core/coap/coap.cpp [hunk 3]
```diff
@@ -646,10 +644,7 @@ void Coap::ProcessReceivedRequest(Header &aHeader, Message &aMessage, const Ip6:
     Message *cachedResponse = NULL;
     otError error = OT_ERROR_NOT_FOUND;
 
-    if (mInterceptor != NULL)
-    {
-        SuccessOrExit(error = mInterceptor(aMessage, aMessageInfo, mContext));
-    }
+    SuccessOrExit(GetNetif().ApplyTmfFilter(aMessageInfo, GetNetif().GetMle()));
```
Comment (bukepo) [inline src/core/coap/coap.cpp]: I prefer keeping the `Interceptor` for it don't strengthen CoAP's dependence on `ThreadNetif`.

File: src/core/net/ip6.cpp [hunk 4]
```diff
@@ -594,17 +592,30 @@ otError Ip6::ProcessReceiveCallback(const Message &aMessage, const MessageInfo &
...
 
         case kProtoUdp:
-            if (aMessageInfo.GetSockAddr().IsLinkLocal() ||
-                aMessageInfo.GetSockAddr().IsLinkLocalMulticast())
+        {
+            UdpHeader udp;
+            aMessage.Read(aMessage.GetOffset(), sizeof(udp), &udp);
+
+            // do not pass MLE messages
+            if (udp.GetDestinationPort() == Mle::kUdpPort)
```
Comment (bukepo) [inline src/core/net/ip6.cpp]: Use `switch`?

**Pull request merged**